From 31e64b8ae882745b43b7ba70c873d491201a108b Mon Sep 17 00:00:00 2001
From: Mathavan Magesh <mmagesh@maxlinear.com>
Date: Wed, 28 Aug 2024 13:51:17 +0800
Subject: [PATCH] WLANRTSYS-84216 Reject ML Association request without EHT
 capability in sibling STA

Issue:
UMI_STA_ADD is received in firmware with ML capability and without EHT
capability for sibling STA. FW asserted while setting rate for the station as
the rateMask is zero.

Fix:
Reject Association in hostapd if STA sends Assoc Req with ML IE
and without EHT capability in sibling STA.

FIXUP to WLANRTSYS-62845 MLO changes: MLO Auth Assoc, WPA3 and eapol
changes
---
 src/ap/ieee802_11.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 59909d42b..1a805b989 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -5660,6 +5660,7 @@ static void handle_assoc(struct hostapd_data *hapd,
 	struct sta_info *sta;
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	struct sta_info *unlinked_sta;
+	struct sta_info *linked_sta;
 #endif
 	u8 *tmp = NULL;
 #ifdef CONFIG_FILS
@@ -5925,7 +5926,7 @@ static void handle_assoc(struct hostapd_data *hapd,
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	/*If the assoc req has ML IE without EHT capability, reject the association.*/
-	if((sta->mxl_sta_info.ext_flags & WLAN_STA_MULTILINK) && !(sta->flags & WLAN_STA_EHT)) {
+	if ((sta->mxl_sta_info.ext_flags & WLAN_STA_MULTILINK) && !(sta->flags & WLAN_STA_EHT)) {
 		wpa_printf(MSG_DEBUG, "Reject ML STA " MACSTR " Assoc Request (missing EHT capability)", MAC2STR(mgmt->sa));
 		resp = WLAN_STATUS_UNSPECIFIED_FAILURE;
 		goto fail;
@@ -5943,6 +5944,12 @@ static void handle_assoc(struct hostapd_data *hapd,
 		}
 	}
 
+	linked_sta = sta->mxl_sta_info.linked_sta;
+	if (linked_sta && !(linked_sta->flags & WLAN_STA_EHT) && (linked_sta->mxl_sta_info.ext_flags & WLAN_STA_MULTILINK)) {
+		wpa_printf(MSG_DEBUG, "Reject ML STA " MACSTR " Assoc Request (missing EHT capability) for sibling STA", MAC2STR(mgmt->sa));
+		resp = WLAN_STATUS_UNSPECIFIED_FAILURE;
+		goto fail;
+	}
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 #ifdef CONFIG_VENDOR_MXL
 	if (!mxl_hostapd_check_softblock_connect(hapd, mgmt->sa, &resp))
-- 
2.43.0

